rfd/0175-spiffe-federation.md

```diff
@@ -0,0 +1,429 @@
+// from a HTTPS endpoint that is protected by a Web PKI certificate.
+message SPIFFEFederationBundleSourceHTTPSWeb {
+  // The URL of the SPIFFE Bundle Endpoint.
+  string bundle_endpoint_url = 1;
+}
+
+// SPIFFEFederationBundleSourceHTTPSSPIFFE is a bundle source that fetches the bundle
+// from a HTTPS endpoint that is protected by a SPIFFE certificate that is
+// "self-served" (i.e. the SPIFFE certificate is issued by the same trust domain
+// that the bundle is for).
+message SPIFFEFederationBundleSourceHTTPSSPIFFE {
+  // The URL of the SPIFFE Bundle Endpoint.
+  string bundle_endpoint_url = 1;
+  // The initial SPIFFE bundle that is used to bootstrap the connection to the
+  // bundle endpoint. After the first sync, this field will no longer be used.
+  string bundle_bootstrap = 2;
+}
+
+// SPIFFEFederationBundleSource configures how the federation bundle is sourced.
+// Only one field can be set.
+message SPIFFEFederationBundleSource {
+  SPIFFEFederationBundleSourceStatic static = 1;
+  SPIFFEFederationBundleSourceHTTPSWeb https_web = 2;
+  SPIFFEFederationBundleSourceHTTPSSPIFFE https_spiffe = 3;
+}
+
+// SPIFFEFederationSpec is the configuration of a trust domain federation.
+message SPIFFEFederationSpec {
+  // The source of the federation bundle.
+  SPIFFEFederationBundleSource bundle_source = 1;
+}
+
+// FederationStatus is the status of a trust domain federation.
+message SPIFFEFederationStatus {
+  // The most recently fetched bundle from the federated trust domain.
+  string current_bundle = 1;
+  // The time that the most recently fetched bundle was obtained.
+  google.protobuf.Timestamp current_bundle_synced_at = 2;
+  // The duration that the current bundle suggests the next bundle should be 
+  // refresh after.
+  google.protobuf.Duration current_bundle_refresh_hint = 3;
+}
+```
+
+The `SPIFFEFederation` resource will be accessed via the standard RFD153
+gRPC CRUD APIs, and will be guarded by RBAC. We will grant the verbs to view,
+modify and create `SPIFFEFederation` resources to the default `edit` role.
+
+For security reasons, we will guard the `status` field of the `SPIFFEFederation`
+resource from modification by the user. This field will be set by the Teleport
+Auth Server and there is no need for the user to modify it.
+
+#### Syncing Trust Bundles.
+
+A background task will be introduced to the Teleport Auth Server:
+`sync-spiffe-federation`. This will run on an elected Teleport Auth Server,
+and periodically fetch the trust bundle from the configured Bundle Endpoint.
+This will be written into `status.current_bundle`.
+
+The `current_bundle_refresh_hint` and `current_bundle_synced_at` fields will be
+used to determine when the next fetch should occur. If the `current_bundle` is
+older than the `current_bundle_refresh_hint` from the `current_bundle_synced_at`,
+a new fetch will be initiated.  If unspecified, it will be assumed that the
+bundle should be refreshed every 60 minutes.
```

Comment:
I think 60 minutes is probably also a perfectly sane value, though the spec does recommend 5 minutes when unspecified: https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE_Federation.md#41-adding-and-removing-keys

> Clients SHOULD poll at a frequency equal to the value of the bundle’s `spiffe_refresh_hint`, in seconds. If not set, a reasonably low default value should apply - five minutes is recommended.

Reply:
Good catch - I think 5 minutes is perfectly fine and is recommended by the SPIFFE spec. I'll adjust and make reference to the spec.

Reply:
ca8056b66a